File: mcmc/ruby/mcmc.rb
# Target probability distribution: a mixture of two Gaussian distributions
def target_distribution(x)
    gaussian1 = 0.3 * Math.exp(-0.2 * (x - 10)**2)
    gaussian2 = 0.7 * Math.exp(-0.2 * (x + 10)**2)
    gaussian1 + gaussian2
  end
  
  # Metropolis-Hastings algorithm for MCMC sampling
  def metropolis_hastings(target, num_samples, initial_state, proposal_std)
    current_state = initial_state
    accepted = 0
    samples = []
  
    num_samples.times do
      # Generate a proposal state from a normal distribution
      proposal = current_state + proposal_std * rand * 2 - proposal_std
  
      # Calculate the acceptance probability
      acceptance_prob = [1, target.call(proposal) / target.call(current_state)].min
  
      # Accept or reject the proposal
      if rand < acceptance_prob
        current_state = proposal
        accepted += 1
      end
  
      # Save the current state as a sample
      samples << current_state
    end
  
    acceptance_rate = accepted.to_f / num_samples
    puts "Acceptance rate: #{acceptance_rate.round(4)}"
    samples
  end
  
  # Define parameters for MCMC sampling
  num_samples = 10000
  initial_state = 0.0
  proposal_std = 5.0
  
  # Run MCMC sampling using Metropolis-Hastings algorithm
  samples = metropolis_hastings(method(:target_distribution), num_samples, initial_state, proposal_std)
  
  # Print the sampled distribution
  samples.each { |sample| puts sample }
  

File: mcmc/r/mcmc.r
# Target probability distribution: a mixture of two Gaussian distributions
targetDistribution <- function(x) {
  gaussian1 <- 0.3 * exp(-0.2 * (x - 10)^2)
  gaussian2 <- 0.7 * exp(-0.2 * (x + 10)^2)
  return(gaussian1 + gaussian2)
}

# Metropolis-Hastings algorithm for MCMC sampling
metropolisHastings <- function(target, numSamples, initialState, proposalStd) {
  currentState <- initialState
  accepted <- 0
  samples <- numeric(numSamples)
  
  for (i in 1:numSamples) {
    # Generate a proposal state from a normal distribution
    proposal <- currentState + proposalStd * rnorm(1)
    
    # Calculate the acceptance probability
    acceptanceProb <- min(1, target(proposal) / target(currentState))
    
    # Accept or reject the proposal
    if (runif(1) < acceptanceProb) {
      currentState <- proposal
      accepted <- accepted + 1
    }
    
    # Save the current state as a sample
    samples[i] <- currentState
  }
  
  acceptanceRate <- accepted / numSamples
  cat("Acceptance rate:", round(acceptanceRate, 4), "\n")
  return(samples)
}

# Define parameters for MCMC sampling
numSamples <- 10000
initialState <- 0
proposalStd <- 5

# Run MCMC sampling using Metropolis-Hastings algorithm
samples <- metropolisHastings(targetDistribution, numSamples, initialState, proposalStd)

# Print the sampled distribution
for (sample in samples) {
  cat(sample, "\n")
}
